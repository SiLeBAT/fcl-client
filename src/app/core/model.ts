export interface ValidationError {
    code: string;
    message: string;
}
