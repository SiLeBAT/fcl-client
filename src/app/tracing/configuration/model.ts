import {
    Color,
    DataServiceInputState,
    LabelPart,
    NodeShapeType,
    TableColumn,
    TableRow,
} from "../data.model";
import { ComplexFilterCondition, PropToValuesMap } from "./configuration.model";

export interface RowFilter<T> {
    filter(arr: TableRow[]): TableRow[];
    getSettings(): T;
}

export interface TextFilterSettings {
    filterTerm: string;
    filterProps: string[];
}

export enum TableType {
    STATIONS,
    DELIVERIES,
}

export type TreeStatus = "collapsed" | "expanded";

export interface EditHighlightingServiceData {
    favouriteProperties: TableColumn[];
    otherProperties: TableColumn[];
    propToValuesMap: PropToValuesMap;
    ruleListItems: RuleListItem[];
}

export interface RuleListItem {
    id: RuleId;
    name: string;
    color: Color | null;
    shape: NodeShapeType | null;
    showInLegend: boolean;
    disabled: boolean;
    autoDisabled: boolean;
    isAnonymizationRule: boolean;
    effElementsCountTooltip: string;
    effElementsCount: number;
    conflictCount: number;
    ruleType: RuleType | null;
}

export interface EditRuleCore {
    id: string;
    name: string;
    userDisabled: boolean;
    complexFilterConditions: ComplexFilterCondition[];
    isValid: boolean;
}

export type EditRule =
    | ColorAndShapeEditRule
    | ColorEditRule
    | LabelEditRule
    | InvEditRule;

export enum RuleType {
    INVISIBILITY,
    LABEL,
    COLOR_AND_SHAPE,
    COLOR,
}

export interface ColorAndShapeEditRule extends EditRuleCore {
    type: RuleType.COLOR_AND_SHAPE;
    showInLegend: boolean;
    color: Color | null;
    shape: NodeShapeType | null;
}

export interface ColorEditRule extends EditRuleCore {
    type: RuleType.COLOR;
    showInLegend: boolean;
    color: Color;
}

export interface SimpleLabelEditRule extends EditRuleCore {
    type: RuleType.LABEL;
    labelProperty: string | null;
}

export interface ComposedLabelEditRule extends EditRuleCore {
    type: RuleType.LABEL;
    labelParts: LabelPart[];
    labelPrefix: string;
}

export type LabelEditRule = SimpleLabelEditRule | ComposedLabelEditRule;

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface InvEditRule extends EditRuleCore {
    type: RuleType.INVISIBILITY;
}

export type StationEditRule =
    | ColorAndShapeEditRule
    | LabelEditRule
    | InvEditRule;
export type DeliveryEditRule = LabelEditRule | InvEditRule | ColorEditRule;

export type DeliveryRuleType = RuleType.LABEL | RuleType.COLOR;
export type StationRuleType = RuleType.LABEL | RuleType.COLOR_AND_SHAPE;

export type RuleId = string;

export interface EditHighlightingState<T extends EditRule> {
    dataServiceInputState: DataServiceInputState;
    editRules: T[];
}
