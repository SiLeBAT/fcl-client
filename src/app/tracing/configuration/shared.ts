import {
    DeliveryHighlightingRule, HighlightingRule, LinePatternType,
    StationHighlightingRule, TableColumn, TableRow
} from '../data.model';
import { PropToValuesMap } from './configuration.model';
import {
    ColorAndShapeEditRule, ColorEditRule, InvEditRule,
    LabelEditRule, RowFilter, RuleType, SimpleLabelEditRule
} from './model';
import * as _ from 'lodash';

export type EditRuleOfType<T extends RuleType> =
    T extends RuleType.LABEL ? LabelEditRule :
    T extends RuleType.COLOR_AND_SHAPE ? ColorAndShapeEditRule :
    T extends RuleType.COLOR ? ColorEditRule :
    T extends RuleType.INVISIBILITY ? InvEditRule : never;

export function filterTableRows(rows: TableRow[], rowFilters: RowFilter<any>[]): TableRow[] {
    let filteredRows = rows;
    for (const rowFilter of rowFilters) {
        filteredRows = rowFilter.filter(filteredRows);
    }
    return filteredRows;
}

export function extractPropToValuesMap(tableRows: TableRow[], tableColumns: TableColumn[]): PropToValuesMap {
    const propToValuesMap: PropToValuesMap = {};
    for (const column of tableColumns) {
        const values: string[] = _.uniq(
            tableRows
                .map(r => r[column.id] as (string | number | boolean))
                .filter(v => v !== undefined && v !== null)
                .map(v => typeof v === 'string' ? v : '' + v)
        ).sort();
        propToValuesMap[column.id] = values;
    }

    propToValuesMap[''] = _.uniq([].concat(...Object.values(propToValuesMap))).sort();

    return propToValuesMap;
}

export function isSimpleLabelRule(editRule: LabelEditRule): editRule is SimpleLabelEditRule {
    return (editRule as SimpleLabelEditRule).labelProperty !== undefined;
}

export function createDefaultHRule(): Omit<HighlightingRule, 'id' | 'name'> {
    return {
        userDisabled: false,
        autoDisabled: false,
        showInLegend: false,
        color: null,
        invisible: false,
        adjustThickness: false,
        labelProperty: null,
        valueCondition: null,
        logicalConditions: [[]]
    };
}

export function createDefaultStatHRule(): Omit<StationHighlightingRule, 'id' | 'name'> {
    return {
        ...createDefaultHRule(),
        shape: null
    };
}

export function createDefaultDelHRule(): Omit<DeliveryHighlightingRule, 'id' | 'name'> {
    return {
        ...createDefaultHRule(),
        linePattern: LinePatternType.SOLID
    };
}
