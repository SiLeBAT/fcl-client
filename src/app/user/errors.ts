import { InvalidServerInputHttpErrorResponse } from '@app/core/errors';
import { ValidationError } from '@app/core/model';
import { CODE_TO_FIELD_MAP } from './consts/error-code-mappings.consts';
import { RegistrationCredentials } from './models/user.model';

export class InvalidRegistrationInput {
    private _errors: Partial<Record<keyof RegistrationCredentials, ValidationError[]>> = {};
    private _message: string | null = null;

    get errors(): Partial<Record<keyof RegistrationCredentials, ValidationError[]>> {
        return this._errors;
    }

    get message(): string | null {
        return this._message;
    }

    constructor(error: InvalidServerInputHttpErrorResponse) {
        this._message = error.message;
        this.mapErrors(error.errors);
    }

    private mapErrors(validationErrors: ValidationError[]): void {
        this._errors = {};
        validationErrors.forEach(error => {
            const field = CODE_TO_FIELD_MAP[error.code] as undefined | keyof RegistrationCredentials;
            if (field !== undefined) {
                const fieldErrors = this._errors[field];
                if (fieldErrors !== undefined) {
                    fieldErrors.push(error);
                } else {
                    this._errors[field] = [error];
                }
            }
        });
    }
}
