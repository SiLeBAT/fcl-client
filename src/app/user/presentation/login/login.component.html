<form [formGroup]="loginForm" (ngSubmit)="onLogin()">
    <div class="fcl-mat-card-content">
        <div class="form-group fcl-form-email">
            <mat-form-field class="fcl-form-field" appearance="standard">
                <mat-label>Email</mat-label>
                <input matInput formControlName="email" name="email" required fclFocusAfterViewInit/>
                <mat-icon matPrefix>mail</mat-icon>
                <mat-error *ngIf="!validateField('email') && loginForm.controls.email.errors.email">
                    Not a valid email
                </mat-error>
                <mat-error *ngIf="!validateField('email') && loginForm.controls.email.errors.required">
                    You must enter a valid email
                </mat-error>
            </mat-form-field>
        </div>

        <div class="form-group">
            <mat-form-field class="fcl-form-field" appearance="standard">
                <mat-label>Password</mat-label>
                <input matInput formControlName="password" type="password" name="password" required />
                <mat-icon matPrefix>lock</mat-icon>
                <mat-error *ngIf="!validateField('password') && loginForm.controls.password.errors.required">
                    Password is required
                </mat-error>
            </mat-form-field>
        </div>

        <div class="fcl-card-action fcl-card-action-container">
            <div class="form-group">
                <button class="fcl-login-form__submit-button" mat-raised-button color="primary" type="submit" [disabled]="!loginForm.valid">
                    Login
                </button>
            </div>
        </div>
    </div>
</form>
