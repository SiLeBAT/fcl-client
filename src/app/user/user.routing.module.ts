import { NgModule } from "@angular/core";
import { Routes, RouterModule, mapToCanActivate } from "@angular/router";

import { LoginViewComponent } from "./presentation/login-view/login-view.component";
import { RegisterViewComponent } from "./presentation/register-view/register-view.component";
import { RecoveryViewComponent } from "./presentation/recovery-view/recovery-view.component";
import { ResetViewComponent } from "./presentation/reset-view/reset-view.component";
import { ActivateViewComponent } from "./presentation/activate-view/activate-view.component";
import { AdminActivateViewComponent } from "./presentation/admin-activate-view/admin-activate-view.component";
import { ProfileContainerComponent } from "./container/profile-container/profile-container.component";
import { environment } from "./../../environments/environment";
import { AuthGuard } from "./guards/auth.guard";
import { NewsActivateViewComponent } from "./presentation/news-activate-view/news-activate-view.component";

const userRoutes: Routes = [
    {
        path: "users",
        children: [
            {
                path: "profile",
                component: ProfileContainerComponent,
                canActivate: environment.serverless
                    ? undefined
                    : mapToCanActivate([AuthGuard]),
            },
            {
                path: "login",
                component: LoginViewComponent,
            },
            {
                path: "register",
                component: RegisterViewComponent,
            },
            {
                path: "recovery",
                component: RecoveryViewComponent,
            },
            {
                path: "reset/:id",
                component: ResetViewComponent,
            },
            {
                path: "activate/:id",
                component: ActivateViewComponent,
            },
            {
                path: "adminactivate/:id",
                component: AdminActivateViewComponent,
            },
            {
                path: "newsactivate/:id",
                component: NewsActivateViewComponent,
            },
        ],
    },
];

@NgModule({
    imports: [RouterModule.forChild(userRoutes)],
    exports: [RouterModule],
})
export class UserRoutingModule {}
