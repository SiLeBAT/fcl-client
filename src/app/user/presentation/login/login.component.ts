import { Component, OnInit, Output, EventEmitter } from '@angular/core';
import { Validators, FormControl, FormGroup } from '@angular/forms';

import { LoginCredentials } from '../../models/user.model';

interface LoginForm {
    email: FormControl<string | null>;
    password: FormControl<string | null>;
}

@Component({
    selector: 'fcl-login',
    templateUrl: './login.component.html',
    styleUrls: ['./login.component.scss']
})
export class LoginComponent implements OnInit {
    @Output() login = new EventEmitter<LoginCredentials>();
    loginForm: FormGroup<LoginForm>;

    ngOnInit() {
        this.loginForm = new FormGroup<LoginForm>({
            email: new FormControl<string | null>(null, [Validators.required, Validators.email]),
            password: new FormControl<string | null>(null, Validators.required)
        });
    }

    onLogin() {
        const credentials: LoginCredentials = {
            email: this.loginForm.value.email,
            password: this.loginForm.value.password
        };
        this.login.emit(credentials);
        this.loginForm.reset();
    }

    validateField(fieldName: string) {
        return this.loginForm.controls[fieldName].valid
               || this.loginForm.controls[fieldName].untouched;
    }
}
