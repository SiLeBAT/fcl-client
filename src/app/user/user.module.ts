import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { PasswordStrengthMeterModule } from 'angular-password-strength-meter';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';

import { SharedModule } from './../shared/shared.module';
import { CoreModule } from '../core/core.module';
import { MainPageModule } from '../main-page/main-page.module';
import { UserRoutingModule } from './user.routing.module';
import { ContentModule } from '@app/content/content.module';

import { STATE_SLICE_NAME, reducer } from './state/user.reducer';
import { UserEffects } from './state/user.effects';
import { LoginContainerComponent } from './container/login-container/login-container.component';
import { LoginComponent } from './presentation/login/login.component';
import { LoginViewComponent } from './presentation/login-view/login-view.component';
import { RegisterContainerComponent } from './container/register-container/register-container.component';
import { RegisterComponent } from './presentation/register/register.component';
import { RegisterViewComponent } from './presentation/register-view/register-view.component';
import { RecoveryContainerComponent } from './container/recovery-container/recovery-container.component';
import { RecoveryComponent } from './presentation/recovery/recovery.component';
import { RecoveryViewComponent } from './presentation/recovery-view/recovery-view.component';
import { ResetContainerComponent } from './container/reset-container/reset-container.component';
import { ResetComponent } from './presentation/reset/reset.component';
import { ResetViewComponent } from './presentation/reset-view/reset-view.component';
import { ActivateContainerComponent } from './container/activate-container/activate-container.component';
import { ActivateComponent } from './presentation/activate/activate.component';
import { ActivateViewComponent } from './presentation/activate-view/activate-view.component';
import { AdminActivateContainerComponent } from './container/admin-activate-container/admin-activate-container.component';
import { AdminActivateComponent } from './presentation/admin-activate/admin-activate.component';
import { AdminActivateViewComponent } from './presentation/admin-activate-view/admin-activate-view.component';
import { ProfileContainerComponent } from './container/profile-container/profile-container.component';
import { ProfileComponent } from './presentation/profile/profile.component';
import { GdprAgreementComponent } from './presentation/gdpr-agreement/gdpr-agreement.component';
import { NewsActivateContainerComponent } from './container/news-activate-container/news-activate-container.component';
import { NewsActivateComponent } from './presentation/news-activate/news-activate.component';
import { NewsActivateViewComponent } from './presentation/news-activate-view/news-activate-view.component';

@NgModule({
    imports: [
        CommonModule,
        ContentModule,
        ReactiveFormsModule,
        PasswordStrengthMeterModule,
        SharedModule,
        CoreModule,
        MainPageModule,
        UserRoutingModule,
        StoreModule.forFeature(STATE_SLICE_NAME, reducer),
        EffectsModule.forFeature([UserEffects]),
        FormsModule
    ],
    exports: [],
    declarations: [
        LoginContainerComponent,
        LoginComponent,
        LoginViewComponent,
        RegisterContainerComponent,
        RegisterComponent,
        RegisterViewComponent,
        RecoveryViewComponent,
        RecoveryComponent,
        RecoveryContainerComponent,
        ResetContainerComponent,
        ResetComponent,
        ResetViewComponent,
        ActivateContainerComponent,
        ActivateComponent,
        ActivateViewComponent,
        AdminActivateContainerComponent,
        AdminActivateComponent,
        AdminActivateViewComponent,
        ProfileContainerComponent,
        ProfileComponent,
        GdprAgreementComponent,
        NewsActivateContainerComponent,
        NewsActivateComponent,
        NewsActivateViewComponent
    ],
    schemas: [CUSTOM_ELEMENTS_SCHEMA]

})
export class UserModule { }
