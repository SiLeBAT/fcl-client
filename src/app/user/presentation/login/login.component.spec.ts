import {async, ComponentFixture, TestBed} from '@angular/core/testing';

import {LoginComponent} from './login.component';
import {MaterialModule} from '../../../shared/material.module';
import {SharedModule} from '../../../shared/shared.module';
import {LoginCredentials} from '../../models/user.model';

describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;

  beforeEach(async(() => {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    TestBed.configureTestingModule({
      declarations: [LoginComponent],
      imports: [MaterialModule, SharedModule],
    })
      .compileComponents()
      .then(() => {
        fixture = TestBed.createComponent(LoginComponent);
        component = fixture.componentInstance;
      });
  }));

  it('should create the component', () => {
    expect(component).toBeTruthy();
  });

  it('should render the component correctly', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should have an invalid login form when empty', () => {
    expect(component.loginForm.valid).toBeFalsy();
  });

  it('should have an invalid email field when being empty', () => {
    const emailField = component.loginForm.controls['email'];

    expect(emailField.valid).toBeFalsy();
  });

  it('should have an required validator on the email field', () => {
    let errors = {};
    const emailField = component.loginForm.controls['email'];
    errors = emailField.errors || {};

    expect(errors['required']).toBeTruthy();
  });

  it('should have an invalid email field with falsy email value', () => {
    const emailField = component.loginForm.controls['email'];
    emailField.setValue('test');
    const errors = emailField.errors || {};

    expect(errors['required']).toBeFalsy();
    expect(errors['email']).toBeTruthy();
  });

  it('should have a valid email field with valid email value', () => {
    const emailField = component.loginForm.controls['email'];
    emailField.setValue('test@test.com');
    const errors = emailField.errors || {};

    expect(errors['required']).toBeFalsy();
    expect(errors['email']).toBeFalsy();
  });

  it('should have an invalid password field when being empty', () => {
    const passwordField = component.loginForm.controls['password'];

    expect(passwordField.valid).toBeFalsy();
  });

  it('should have an required validator on the password field', () => {
    let errors = {};
    const passwordField = component.loginForm.controls['password'];
    errors = passwordField.errors || {};

    expect(errors['required']).toBeTruthy();
  });

  it('should have a valid password field with password value', () => {
    const passwordField = component.loginForm.controls['password'];
    passwordField.setValue('test');
    const errors = passwordField.errors || {};

    expect(errors['required']).toBeFalsy();
  });

  it('should emit the login credentials when submitting the login form', () => {
    const email = 'test@test.com';
    const password = 'test';

    expect(component.loginForm.valid).toBeFalsy();

    component.loginForm.controls['email'].setValue(email);
    component.loginForm.controls['password'].setValue(password);

    expect(component.loginForm.valid).toBeTruthy();

    let credentials: LoginCredentials;
    component.login.subscribe(value => (credentials = value));

    component.onLogin();

    expect(credentials!.email).toBe(email);
    expect(credentials!.password).toBe(password);
  });
});
