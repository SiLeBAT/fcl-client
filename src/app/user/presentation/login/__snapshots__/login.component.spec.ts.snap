// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should render the component correctly 1`] = `
<fcl-login
  login={[Function EventEmitter_]}
  loginForm={[Function FormGroup]}
>
  <form
    novalidate=""
  >
    <div
      class="fcl-mat-card-content"
    >
      <div
        class="form-group fcl-form-email"
      >
        <mat-form-field
          appearance="standard"
          class="fcl-form-field mat-form-field mat-primary"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  class="mat-input-element mat-form-field-autofill-control"
                  formcontrolname="email"
                  matinput=""
                  name="email"
                  required=""
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              
            </div>
          </div>
        </mat-form-field>
      </div>
      <div
        class="form-group"
      >
        <mat-form-field
          appearance="standard"
          class="fcl-form-field mat-form-field mat-primary"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  class="mat-input-element mat-form-field-autofill-control"
                  formcontrolname="password"
                  matinput=""
                  name="password"
                  required=""
                  type="password"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              
            </div>
          </div>
        </mat-form-field>
      </div>
      <div
        class="fcl-card-action fcl-card-action-container"
      >
        <div
          class="form-group"
        >
          <button
            class="fcl-login-form__submit-button mat-focus-indicator mat-raised-button mat-button-base"
            color="primary"
            mat-raised-button=""
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
               Login 
            </span>
            <span
              class="mat-button-ripple mat-ripple"
              matripple=""
            />
            <span
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
</fcl-login>
`;
